src/py/flwr/client/process/clientappio_servicer.py

```diff
@@ -0,0 +1,111 @@
+
+from flwr.common import Context, Message, typing
+from flwr.common.logger import log
+from flwr.common.serde import (
+    clientappstatus_to_proto,
+    context_from_proto,
+    context_to_proto,
+    message_from_proto,
+    message_to_proto,
+    run_to_proto,
+)
+from flwr.common.typing import Run
+
+# pylint: disable=E0611
+from flwr.proto import clientappio_pb2_grpc
+from flwr.proto.clientappio_pb2 import (  # pylint: disable=E0401
+    PullClientAppInputsRequest,
+    PullClientAppInputsResponse,
+    PushClientAppOutputsRequest,
+    PushClientAppOutputsResponse,
+)
+from flwr.proto.message_pb2 import Context as ProtoContext
+from flwr.proto.message_pb2 import Message as ProtoMessage
+from flwr.proto.run_pb2 import Run as ProtoRun
+
+
+# pylint: disable=C0103,W0613,W0201
+class ClientAppIoServicer(clientappio_pb2_grpc.ClientAppIoServicer):
+    """ClientAppIo API servicer."""
+
+    def set_object(
+        self,
+        message: Message,
+        context: Context,
+        run: Run,
+        token: int,
+    ) -> None:
+        """Set client app objects."""
+        log(DEBUG, "ClientAppIo.SetObject")
+        # Serialize Message, Context, and Run
+        self.proto_message: ProtoMessage = message_to_proto(message)
+        self.proto_context: ProtoContext = context_to_proto(context)
+        self.proto_run: ProtoRun = run_to_proto(run)
+        self.token: int = token
+
+    def get_object(self) -> Tuple[Message, Context]:
+        """Get client app objects."""
+        log(DEBUG, "ClientAppIo.GetObject")
+        return self.message, self.context
+
+    def _update_object(self) -> None:
+        """Update client app objects."""
+        log(DEBUG, "ClientAppIo.UpdateObject")
+        # Deserialize Message and Context
+        self.message: Message = message_from_proto(self.proto_message)
+        self.context: Context = context_from_proto(self.proto_context)
+
+    def PullClientAppInputs(
+        self, request: PullClientAppInputsRequest, context: grpc.ServicerContext
+    ) -> PullClientAppInputsResponse:
+        """Pull Message, Context, and Run."""
+        log(DEBUG, "ClientAppIo.PullInputs")
+        if request.token != self.token:
+            raise ValueError("Mismatch between ClientApp and SuperNode token")
+        return PullClientAppInputsResponse(
+            message=self.proto_message,
+            context=self.proto_context,
+            run=self.proto_run,
+        )
+
+    def PushClientAppOutputs(
+        self, request: PushClientAppOutputsRequest, context: grpc.ServicerContext
+    ) -> PushClientAppOutputsResponse:
+        """Push Message and Context."""
+        log(DEBUG, "ClientAppIo.PushOutputs")
+        if request.token != self.token:
+            raise ValueError("Mismatch between ClientApp and SuperNode token")
+        self.proto_message = request.message
+        self.proto_context = request.context
+        # Update Message and Context
+        try:
+            self._update_object()
```

Comment:
I see a bit ahead in #3978, this will be the mechanism by which the `SuperNode` will access the returned `Message/Context` by the `ClientApp`? is this the main functionality behind this method? I wonder if it could be renamed to something other than `get/update_object()`. I thought of `_payload()` (in the end we are sending several objects) but probably are better alternatives.

Reply:
Yes, it happens in `client/app.py`. `_payload()` also works and is more specific.